dprint/dprint
# feat: Wasm plugins v4

Slowly starting work on this. Will implement everything from https://github.com/dprint/dprint/issues/532

Wasm:

- [x] Wasm schema should not require instantiating a wasm module (it should be figured out by just analyzing the import names)
- [x] More efficient communication (no buffering)
- [x] updates from https://github.com/dprint/dprint/issues/714
- [x] updates from https://github.com/dprint/dprint/issues/716
- [x] Cancellation
- [x] Multiple configs
- [x] Range formatting
  - [x] Host range formatting
- [x] ~~Ability for plugins to tell if no_color and if a tty. Probably send on init~~
  - Edit: Nah. CLI can just strip the
- [x] ~~Ability for plugins to do comparisons with language aware algorithms (ex. for sorting imports).~~
  - Edit: No, too much complexity.

Wasm and Process plugins:

- [x] ~~Send log level on initialize so plugins can output with verbose logging.~~
  - ~~Maybe this should be a general init function that also sends info like what the host is (ex. a string like "dprint x.x.x")~~
  - Edit: This isn't so useful I think and might cause a lot of spam in the verbose logs. I think this should be set as config on each plugin.
- [x] Send past version with upgrade message (if not already done)
- [x] ~~Have ability to get information from the host about other plugins via the config id https://github.com/dprint/dprint/issues/532#issuecomment-2172747545~~ (edit: seems not worth the complexity for now)

Todo:

- [x] Update js-formatter before merging to implement this - https://github.com/dprint/js-formatter/pull/12

Closes #532
